### PR TITLE
#2610 signal duplicate properties while parsing

### PR DESCRIPTION
**Pull request by Kasdejong**
## Description
JSON deserializer should now correctly identify and report duplicate property keys

## Related issues
Resolves issue #2610.

## Testing
Unit tests added to assert (fatal) duplicate property keys errors and to assert their differentiation from (recoverable) duplicate array key errors.

### CODE REVIEW

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs [hunk 1]
```diff
@@ -367,11 +367,28 @@ FhirJsonPocoDeserializerState state
...
                     state.Errors.Add(ERR.USE_OF_UNDERSCORE_ILLEGAL(ref reader, state.Path.GetInstancePath(), propertyMapping.Name, propertyName));
-
+                
                 // Note that repeating simple elements (like Extension.url) do not currently exist in the FHIR serialization
-                result = propertyMapping.IsCollection
-                    ? deserializeNormalList((IList)existingValue!, propertyValueMapping, ref reader, propertyMapping, state)
-                    : deserializeSingleValue(ref reader, propertyValueMapping, propertyMapping, state);
+                if (propertyMapping.IsCollection)
+                {
+                    IList l = (existingValue as IList)!;
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:374]:
```suggestion
                    var l = (IList)existingValue;
```

File: src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs [hunk 2]
```diff
@@ -367,11 +367,28 @@ FhirJsonPocoDeserializerState state
...
+                    if (l.Count > 0)
+                    {
+                        state.Path.IncrementIndex(l.Count);
+                        state.Errors.Add(ERR.DUPLICATE_ARRAY(ref reader, state.Path.GetInstancePath()));
+                    }
+                    result = deserializeNormalList(l, propertyValueMapping, ref reader, propertyMapping, state);
+                } 
+                else 
+                {
+                    // if the property already has a mapping, its key must have been encountered before
```
Comment (ewoutkramer) [inline src/Hl7.Fhir.Base/Serialization/BaseFhirJsonPocoDeserializer.cs:385]:
```suggestion
                    // if the property already has a value, its key must have been encountered before
```

**Comment by ewoutkramer** [comment]
We broke it with our last change! It did work before. So maybe the cast can result in null after all?

**Comment by ewoutkramer** [comment]
> Looks like this should be non-null asserted after all

No, that does not make sense, as the `!` is just a compiler hint for static analysis and has no effect on the generated code.

**Comment by Kasdejong** [comment]
With the compiler hint, the code works though... Could it be a runtime type error (nullable/non-nullable)?

**Review by ewoutkramer — APPROVED** [review]

**Pull request merged**